modules/openxAnalyticsAdapter.js

```diff
@@ -634,7 +667,405 @@ adapterManager.registerAnalyticsAdapter({
+bidderRequests: (2) [{…}, {…}]
+noBids: []
+bidsReceived: []
+winningBids: []
+timeout: 3000
+config: {publisherPlatformId: "a3aece0c-9e80-4316-8deb-faf804779bd1", publisherAccountId: 537143056, sampling: 1, testPipeline: true}
+ */
+function onAuctionInit({auctionId, timestamp: startTime, timeout, adUnitCodes}) {
+  auctionMap[auctionId] = {
+    id: auctionId,
+    startTime,
+    timeout,
+    auctionOrder,
+    adUnitCodesCount: adUnitCodes.length,
+    adunitCodesRenderedCount: 0,
+    auctionCompleted: false,
+    auctionSendDelayTimer: void (0),
+  };
+
+  // setup adunit properties in map
+  auctionMap[auctionId].adUnitCodeToBidderRequestMap = adUnitCodes.reduce((obj, adunitCode) => {
+    obj[adunitCode] = {};
+    return obj;
+  }, {});
+
+  auctionOrder++;
+}
+
+// TODO: type BidRequest
+function onBidRequested(bidRequest) {
+  const {auctionId, auctionStart, refererInfo, bids: bidderRequests, start} = bidRequest;
+  const auction = auctionMap[auctionId];
+  const adUnitCodeToBidderRequestMap = auction.adUnitCodeToBidderRequestMap;
+
+  bidderRequests.forEach(bidderRequest => {
+    const { adUnitCode, bidder, bidId: requestId, mediaTypes, params, src } = bidderRequest;
+
+    adUnitCodeToBidderRequestMap[adUnitCode][requestId] = {
+      bidder,
+      params,
+      mediaTypes,
+      source: src,
+      startTime: start,
+      timedOut: false,
+      bids: {}
+    };
+  });
+}
+
+/**
+ *
+ * @param {BidResponse} bidResponse
+ */
+function onBidResponse(bidResponse) {
+  let {
+    auctionId,
+    adUnitCode,
+    requestId,
+    cpm,
+    creativeId,
+    requestTimestamp,
+    responseTimestamp,
+    ts,
+    mediaType,
+    dealId,
+    ttl,
+    netRevenue,
+    currency,
+    originalCpm,
+    originalCurrency,
+    width,
+    height,
+    timeToRespond: latency,
+    adId,
+    meta
+  } = bidResponse;
+
+  auctionMap[auctionId].adUnitCodeToBidderRequestMap[adUnitCode][requestId].bids[adId] = {
+    cpm,
+    creativeId,
+    requestTimestamp,
+    responseTimestamp,
+    ts,
+    adId,
+    meta,
+    mediaType,
+    dealId,
+    ttl,
+    netRevenue,
+    currency,
+    originalCpm,
+    originalCurrency,
+    width,
+    height,
+    latency,
+    winner: false,
+    rendered: false,
+    renderTime: 0,
+  };
+}
+
+function onBidTimeout(args) {
+  utils
+    ._map(args, value => value)
+    .forEach(({ auctionId, adUnitCode, bidId: requestId }) => {
+      auctionMap[auctionId].adUnitCodeToBidderRequestMap[adUnitCode][requestId].timedOut = true;
+    });
+}
+
+/**
+ *
+ * @param {BidResponse} bidResponse
+ */
+function onBidWon(bidResponse) {
+  const { auctionId, adUnitCode, requestId, adId } = bidResponse;
+  auctionMap[auctionId].adUnitCodeToBidderRequestMap[adUnitCode][requestId].bids[adId].winner = true;
+}
+
+/**
+ *
+ * @param {GoogleTagSlot} slot
+ * @param {string} serviceName
+ */
+function onSlotLoadedV2({ slot }) {
+  const renderTime = Date.now();
+  const auction = getAuctionByGoogleTagSLot(slot);
```

Comment:
Does this return a copy of the auction object, or the actual auction object? 

Should we mutate the auction object?

Reply:
The tracked parts of the auction object.